src/webgpu/api/validation/vertex_state.spec.ts

```diff
@@ -488,183 +459,202 @@ g.test('vertex_shader_input_location_in_vertex_state')
-    t.device.createRenderPipeline(descriptor);
-  }
-});
+    ]);
 
-g.test('check_attribute_offset_out_of_bounds').fn(async t => {
-  const vertexState = {
-    vertexBuffers: [
+    const requiredBaseType = {
+      sint: 'i32',
+      uint: 'u32',
+      snorm: 'f32',
+      unorm: 'f32',
+      float: 'f32',
+    }[kVertexFormatInfo[format].type];
+
+    const success = requiredBaseType === shaderBaseType;
+    t.testVertexState(
+      success,
       {
-        arrayStride: 0,
-        attributes: [
+        vertexBuffers: [
           {
-            format: 'float32x2' as const,
-            offset: kMaxVertexBufferArrayStride - 2 * SIZEOF_FLOAT,
-            shaderLocation: 0,
+            arrayStride: 0,
+            attributes: [{ offset: 0, shaderLocation: 0, format }],
           },
         ],
       },
-    ],
-  };
-  {
-    // Control case, setting max attribute offset to MAX_VERTEX_BUFFER_END - 8
-    const descriptor = t.getDescriptor(vertexState, VERTEX_SHADER_CODE_WITH_NO_INPUT);
-    t.device.createRenderPipeline(descriptor);
-  }
-  {
-    // Control case, setting attribute offset to 8
-    vertexState.vertexBuffers[0].attributes[0].offset = 8;
-    const descriptor = t.getDescriptor(vertexState, VERTEX_SHADER_CODE_WITH_NO_INPUT);
-    t.device.createRenderPipeline(descriptor);
-  }
-  {
-    // Test attribute offset out of bounds
-    vertexState.vertexBuffers[0].attributes[0].offset = kMaxVertexBufferArrayStride - 4;
-    const descriptor = t.getDescriptor(vertexState, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+      shader
+    );
+  });
 
-    t.expectValidationError(() => {
-      t.device.createRenderPipeline(descriptor);
-    });
-  }
-});
+g.test('vertex_attribute_offset_alignment')
+  .desc(
+    `
+    Test that vertex attribute offsets must be aligned to the format's component byte size.
+    - Test for all formats.
+    - Test for various arrayStrides and offsets within that stride
+    - Test for various vertex buffer indices
+    - Test for various amounts of attributes in that vertex buffer`
+  )
+  .cases(
+    params()
+      .combine(poptions('format', kVertexFormats))
+      .combine(poptions('arrayStride', [256, kMaxVertexBufferArrayStride]))
+      .expand(p => {
+        // Compute a bunch of test offsets to test.
+        const { bytesPerComponent, componentCount } = kVertexFormatInfo[p.format];
+        const formatSize = bytesPerComponent * componentCount;
+        const offsetsToTest = [0, bytesPerComponent, p.arrayStride - formatSize];
+
+        // Avoid adding duplicate offsets because the harness will assert each case is different.
+        const halfAlignment = Math.floor(bytesPerComponent / 2);
+        if (halfAlignment !== 0) {
+          offsetsToTest.push(halfAlignment);
+          offsetsToTest.push(p.arrayStride - formatSize - halfAlignment);
+        }
+
+        return poptions('offset', offsetsToTest);
```

Comment:
nit: Could do something like
```ts
        return poptions(
          'offset',
          new Set([
            0,
            halfAlignment,
            bytesPerComponent,
            p.arrayStride - formatSize,
            p.arrayStride - formatSize - halfAlignment,
          ])
        );
```
I think.

OTOH, you could filter out formats with bytesPerComponent of 1. Aren't they kind of pointless for this test?

Reply:
Went with the Set approach.